File: fsharp/src/DarkenColor.fs
module DarkenColor

open System.Text.RegularExpressions

type RGBType = { r: int16; g: int16; b: int16 }

let hex2rgb hex =
    let m = Regex.Match(hex, "^#?([a-f0-9]{2})([a-f0-9]{2})([a-f0-9]{2})$")
    if m.Success then
        m.Groups
        |> Seq.cast<Group>
        |> Seq.skip 1 // zero capture group is always the full string, when it matches
        |> Seq.map (fun m -> m.Value)
        |> Seq.map (fun x -> System.Convert.ToInt16(x, 16))
        |> Seq.toList
        |> (function
            | r :: g :: b :: [] -> Some { r = r; g = g; b = b }
            | _ -> None)
    else None

File: dart/src/darken_color.dart
class RGB {
  int r;
  int g;
  int b;

  RGB(this.r, this.g, this.b);
}

class DarkenColor {
  static RGB hex2rgb(String s) {
    RegExp re = new RegExp(r"^#?([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})$");

    Iterable<RegExpMatch> matches = re.allMatches(s);

    if (matches.length < 1) {
      return null;
    }

    RegExpMatch match = matches.first;

    String rStr = match.group(1);
    String gStr = match.group(2);
    String bStr = match.group(3);

    int r = int.tryParse(rStr, radix: 16);
    int g = int.tryParse(gStr, radix: 16);
    int b = int.tryParse(bStr, radix: 16);

    return new RGB(r, g, b);
  }
}
